File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        lo, hi = 1, position[-1] - position[0]
        res = hi
        def check(dis):
            placed = 1
            last_pos = position[0]
            for p in position:
                if p - last_pos >= dis:
                    placed += 1
                    last_pos = p
            return placed >= m
        while lo <= hi:
            mid = (lo + hi) // 2
            if check(mid):
                res= mid
                lo = mid + 1
            else:
                hi = mid - 1
        return res

File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.cpp
class Solution {
public:
    bool check(vector<int>& position, int m, int distance) {
        int placed = 1;
        int last_pos = position[0];
        for (int i = 1; i < position.size(); ++i) {
            if (position[i] - last_pos >= distance) {
                placed += 1;
                last_pos = position[i];
                if (placed == m) {
                    return true;
                }
            }
        }
        return placed >= m;
    }
    int maxDistance(vector<int>& position, int m) {
        std::sort(position.begin(), position.end());
        int low = 1, high = position.back(), mid, ans = high;
        while (low <= high) {
            mid = (low + high) / 2;
            if (check(position, m, mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return ans;
    }
};